source/CardGames.py

```diff
@@ -186,17 +269,167 @@ def addCardsToDealer(self, amount=1):
+    # Check if the hand is empty
+      if not hand:
+          return None
+
+    # Find the card with the highest value
+      highest_card = max(hand, key=lambda card: card.value)
+    
+    # Return the highest card
+      return highest_card
+
+
 class Poker:
-  def __init__(self, players: "list[Player]", dealer: Dealer):
-    self.players = players
-    self.dealer = dealer
-    self.pot = 0
+    def __init__(self, players, dealer):
+        self.players = players
+        self.dealer = dealer
+        self.pot = Pot()
+
+    def evaluate_hand(self, hand):
+      values = sorted((card.value for card in hand), reverse=True)
+      suits = [card.suit for card in hand]
+      value_counts = {value: values.count(value) for value in values}
+      is_flush = len(set(suits)) == 1
+      is_straight = len(set(values)) == 5 and max(values) - min(values) == 4
+
+      hand_rankings = {
+          "High Card": 1, "Pair": 2, "Two Pair": 3, "Three of a Kind": 4,
+        "Straight": 5, "Flush": 6, "Full House": 7, "Four of a Kind": 8, "Straight Flush": 9
+      }
+
+    # Check for straight flush
+      if is_straight and is_flush:
+          return (hand_rankings["Straight Flush"], max(values))
+      elif is_flush:
+          return (hand_rankings["Flush"], max(values))
+      elif is_straight:
+          return (hand_rankings["Straight"], max(values))
+      elif 4 in value_counts.values():  # Check for four
+          return (hand_rankings["Four of a Kind"], max(value for value, count in value_counts.items() if count == 4))
+      elif sorted(value_counts.values()) == [2, 3]:  # Check for full house
+          three_val = max(value for value, count in value_counts.items() if count == 3)
+          return (hand_rankings["Full House"], three_val)
+      elif 3 in value_counts.values():  # Check for three
+          return (hand_rankings["Three of a Kind"], max(value for value, count in value_counts.items() if count == 3))
+      elif list(value_counts.values()).count(2) == 2:  # Check for two pairs
+          high_pair = max(value for value, count in value_counts.items() if count == 2)
+          return (hand_rankings["Two Pair"], high_pair)
+      elif 2 in value_counts.values():  # Check for a pair
+          return (hand_rankings["Pair"], max(value for value, count in value_counts.items() if count == 2))
+      else:
+          return (hand_rankings["High Card"], max(values))
+
 
-  # Add all functions pertaining to the game of poker below
-  # Add global variables to the __init__ function above
+    def find_winner(self):
+      best_hand = (0, 0)  # Initialize with lowest possible ranking 
+      winners = []
+
+      for player in self.players:
+          player_hand_rank = self.evaluate_hand(player.hand)
+        
+          if player_hand_rank > best_hand:
+              best_hand = player_hand_rank
+              winners = [player]
+          elif player_hand_rank == best_hand:
+              winners.append(player)
+
+      return winners
+
```

Comment:
The function effectively finds the players with the best hand, as it works with different test cases, including those testing with more than one player to identify the flush. The code is well-structured. Good job!

Reply:
Thanks Erin!